docs/source/extending_mypy.rst

```diff
@@ -236,6 +236,12 @@ module. It is called before semantic analysis. For example, this can
 be used if a library has dependencies that are dynamically loaded
 based on configuration information.
 
+**get_config_data()** can be used if the plugin has some sort of
+per-module configuration that can affect typechecking. In that case,
+when the configuration for a module changes, we want to invalidate
+mypy's cache for that module so that it can be rechecked. This hook,
+which should return data encodable as JSON, allows doing this.
```

Comment:
I think this name may be a bit confusing, I would propose something like `report_config_data()`. Maybe also emphasize more that plugins being normal Python modules can read their custom config files (without this method), but since mypy heavily uses caching, the plugin should report whenever cache for given module should be invalidated because a plugin config changed.